skimage/morphology/rolling_ball.py

```diff
@@ -0,0 +1,148 @@
+    Stanley R. Sternberg (1983) in the paper Biomedical Image Processing [1]_.
+
+    Parameters
+    ----------
+    image : ndarray
+        The image to be filtered.
+    radius : float
+        The radius of the ball that is used as the structuring element.
+    white_background : bool
+        If true, the algorithm separates dark features from a bright
+        background.
+
+    Returns
+    -------
+    filtered_image : ndarray
+        The image with background removes
+    background : ndarray
+        The background that was removed.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from skimage import data
+    >>> from skimage.morphology import rolling_ball
+
+    Subtract a black background from an example image
+
+    >>> image = data.coins()
+    >>> result, bg = rolling_ball(image, radius=100)
+
+    Subtract a white background from an example image
+
+    >>> image = data.page()
+    >>> result, bg = rolling_ball(image, radius=100, white_background=True)
+
+    Notes
+    -----
+    Check out the gallery example of this function (TODO: link)
+
+    References
+    ----------
+    .. [1] Sternberg, Stanley R. "Biomedical image processing." Computer 1
+           (1983): 22-34. :DOI:`10.1109/MC.1983.1654163`
+    """
+
+    if not isinstance(radius, float):
+        try:
+            radius = float(radius)
+        except ValueError:
+            raise ValueError(f"Radius should be float, was {type(radius)}")
+
+    if radius <= 0:
+        raise ValueError(f"Radius must be greater zeros, was {radius}")
+
+    try:
+        white_background = bool(white_background)
+    except ValueError:
+        raise ValueError(
+            f"white_background must be convertible to a boolean, "
+            f"was {type(white_background)}"
+        )
+
+    if not isinstance(input_img, np.ndarray):
+        raise ValueError(
+            f"input_img must be a np.ndarray, "
+            f"was {type(input_img)}"
+        )
+
+    if not issubclass(input_img.dtype.type, np.integer):
+        raise ValueError("Currently only integer images are supported.")
+
+    img = input_img.copy()
+    if white_background:
+        img = invert(img)
+
+    # sagitta assuming the position is where the ball touches
+    # the image umbra
+    L = np.arange(-radius, radius + 1)
+    X, Y = np.meshgrid(L, L)
+    distance = np.sqrt(X ** 2 + Y ** 2)
+    sagitta = radius - np.sqrt(radius ** 2 - distance ** 2)
+    sagitta[np.isnan(sagitta)] = np.Inf
+
+    kernel = np.array((X ** 2 + Y ** 2) <= radius ** 2, dtype=float)
+    kernel[kernel == 0] = np.Inf
+
+    kernel_size = np.array(kernel.shape)
+    img_original = img.copy()
+    img_size = img.shape
+    x = img
+    stride = (1, 1)
+
+    # pad the image
+    padding_amount = kernel_size // 2 * 2
+    half_pad = padding_amount // 2
+    img = np.Inf * np.ones(x.shape + padding_amount, dtype=x.dtype)
+    img[half_pad[0]:-half_pad[0], half_pad[1]:-half_pad[1]] = x
```

Comment:
```suggestion
    padded = np.pad(img, kernel_size // 2, mode='constant', constant_values=np.inf)
```

Reply:
I explicitly choose to not use `np.pad` here. I used it in performance-critical code for a different project earlier this year, and it was MUCH much slower for such a simple const-padding compared to doing it manually.

Reply:
(1) Whoa! Well then we can probably use this implementation elsewhere in this codebase! 😂 Maybe you can make this a function `fast_constant_pad` in `skimage._shared`? But,

(2) it would be quite surprising if pad ends up being the limiting factor here, no?

Reply:
I just tried writing a small script to compare timings for this manual version and `np.pad`. In it, `np.pad` was faster (as it should be). Now I doubt myself, so I will do some more profiling to see if `np.pad` is indeed slow, or if there is a bug somewhere in the other project (the latter is more likely I think).

Reply:
It's entirely possible that NumPy's implementation has improved since you last checked. =)

Reply:
@lagru improved `numpy.pad` quite a bit in NumPy >= 0.17 (numpy/numpy#11358).

On reviewing this I see that I had actually suggested a fast path for constant padding in lagru/numpy#3, but this was not merged at that time and I didn't follow up on it later. We could perhaps revisit that in NumPy.

Reply:
On my system, I tried this implementation vs. `np.pad` for x.shape = (2048, 2048) and kernel.shape = (101, 101).

`np.pad` (from NumPy 1.19.0) took 6.33 ms ± 84 µs while the version proposed here takes 11 ms ± 84.5 µs. 

You can improve the version here a little by using `np.full` instead of `np.Inf * np.ones`, but to get performance equivalent to NumPy, I have to basically replicate what numpy is doing, which is an n-dimensional equivalent of the following 2d code:

```Python
def const_pad(x, kernel_size):
    padding_amount = kernel_size // 2 * 2
    half_pad = padding_amount // 2
    img = np.empty(x.shape + padding_amount, dtype=x.dtype)
    img[half_pad[0]:-half_pad[0], half_pad[1]:-half_pad[1]] = x
    img[:half_pad[0], :] = np.Inf
    img[-half_pad[0]:, :] = np.Inf
    img[half_pad[0]:-half_pad[0], :half_pad[0]] = np.Inf
    img[half_pad[0]:-half_pad[0], -half_pad[0]:] = np.Inf
    return img
```

That `const_pad` function gives nearly identical time to `numpy.pad` for me.

Reply:
I think the best course of action is to refactor it using `np.pad` and (potentially) revisit this at a later stage if it turns out to be the bottleneck of this function as @jni suggested.